### PR TITLE
Get torus working with restarts

### PR DESCRIPTION
**Pull request by brryan**
## PR Summary

We weren't able to properly restart torus simulations with radiation due to the higher-dimensional radiation quantities. This PR fixes that, based on the changes to parthenon in https://github.com/parthenon-hpc-lab/parthenon/pull/781. In this PR this mostly required changing `phoedf` to accommodate the new output format. 

This also adds a maximum optical depth per timestep as an input parameter. The default is `1e5`. In some cases very very large optical depths can cause numerical failures in the source terms. 

## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.
- [x] Merge https://github.com/parthenon-hpc-lab/parthenon/pull/781 into `parthenon`
- [x] Merge https://github.com/lanl/singularity-opac/pull/43 into `singularity-opac`

### CODE REVIEW

File: scripts/python/phoedf.py [hunk 1]
```diff
@@ -92,15 +111,75 @@ def flatten_indices(mu, nu):
     self.Pr = None
     self.tau = None
     self.Tg = None
+    self.ucon = None
+    self.bcon = None
+    self.J = None
+    self.Hcov = None
+    self.E = None
+    self.F = None
+    self.P = None
```
Comment (brryan) [inline scripts/python/phoedf.py:120]: Along with fixing the indexing for the new output shapes here, I added some additional quantities that can be evaluated by `phoedf`

File: src/main.cpp [hunk 2]
```diff
@@ -60,11 +60,12 @@ int main(int argc, char *argv[]) {
...
   // call post-initialization
-  phoebus::PostInitializationModifier(pman.pinput.get(), pman.pmesh.get());
+  phoebus::PostInitializationModifier(pman.pinput.get(), pman.pmesh.get(),
+                                      pman.IsRestart());
 
   // Initialize the driver
-  phoebus::PhoebusDriver driver(pman.pinput.get(), pman.app_input.get(),
-                                pman.pmesh.get());
+  phoebus::PhoebusDriver driver(pman.pinput.get(), pman.app_input.get(), pman.pmesh.get(),
+                                pman.IsRestart());
```
Comment (brryan) [inline src/main.cpp:68]: We need to know if the problem is a restart so we can for example NOT rescale the magnetic field on restarts of the torus problem

File: src/microphysics/opac_phoebus/opac_phoebus.cpp [hunk 3]
```diff
@@ -270,8 +273,11 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
       const Real lTMin = pin->GetOrAddReal("mean_opacity", "ltmin", std::log10(1.e5));
       const Real lTMax = pin->GetOrAddReal("mean_opacity", "ltmax", std::log10(1.e12));
       const int NT = pin->GetOrAddInteger("mean_opacity", "nt", 10);
-      MeanSOpacity mean_opac_host = MeanSOpacityCGS(opacity_host, lRhoMin, lRhoMax, NRho,
-                                                    lTMin, lTMax, NT, YeMin, YeMax, NYe);
+      auto cgs_mean_opacity = MeanSOpacityCGS(opacity_host, lRhoMin, lRhoMax, NRho, lTMin,
+                                              lTMax, NT, YeMin, YeMax, NYe);
+      MeanSOpacity mean_opac_host = MeanNonCGSUnitsS<MeanSOpacityCGS>(
+          std::forward<MeanSOpacityCGS>(cgs_mean_opacity), time_unit, mass_unit,
+          length_unit, temp_unit);
```
Comment (brryan) [inline src/microphysics/opac_phoebus/opac_phoebus.cpp:280]: This fixes a unit issue with mean opacities that was fixed in `singularity-opac`

File: src/phoebus_utils/root_find.hpp [hunk 4]
```diff
@@ -29,9 +29,10 @@ struct RootFind {
     if (ya * yb > 0.0) {
       if (status != nullptr) {
         *status = RootFindStatus::failure;
+      } else {
+        printf("Root not bracketed in find_root. a, b, ya, yb = %g %g %g %g\n", a, b, ya,
```
Comment (brryan) [inline src/phoebus_utils/root_find.hpp:33]: Since in the radiation force solver I start with a relatively small root bracket and then if that fails I try again with a larger bracket for a while, we don't want this `printf` statement to fire unless we're not (potentially) catching rootfind errors elsewhere with the `status` argument.

File: src/radiation/moments.cpp [hunk 5]
```diff
@@ -846,54 +847,46 @@ TaskStatus CalculateGeometricSourceImpl(T *rc, T *rc_src) {
...
+          SPACETIMELOOP2(mu, nu) { con_T[mu][nu] += con_n[mu] * con_n[nu] * E; }
           SPACETIMELOOP(mu) {
-            srcE += con_T[mu][0] * dlnalp[mu];
-            SPACETIMELOOP(nu) {
-              Real Gamma_udd = 0.;
-              SPACETIMELOOP(lam) { Gamma_udd += con_g[0][lam] * Gamma[lam][nu][mu]; }
-              srcE -= con_T[mu][nu] * Gamma_udd;
+            SPACELOOP(ii) {
+              con_T[mu][ii + 1] += con_n[mu] * conF(ii);
+              con_T[ii + 1][mu] += con_n[mu] * conF(ii);
```
Comment (brryan) [inline src/radiation/moments.cpp:869]: The previous evaluation of Tmunu for the radiation was incorrect because it assumed H^mu = Htilde^mu

File: src/radiation/moments_source.cpp [hunk 6]
```diff
@@ -351,12 +353,26 @@ TaskStatus MomentFluidSourceImpl(T *rc, Real dt, bool update_fluid) {
...
+                                             src_rootfind_tol * v(iblock, pT, k, j, i),
+                                             v(iblock, pT, k, j, i), &status);
+            Real rebracketing_fac = 10.;
+            int n_rebracketing_tries = 0;
+            constexpr int MAX_REBRACKETING_TRIES = 4;
+            while (status == root_find::RootFindStatus::failure &&
+                   n_rebracketing_tries < MAX_REBRACKETING_TRIES) {
+              T1 = root_find.regula_falsi(
+                  res, 1.e-1 / rebracketing_fac * v(iblock, pT, k, j, i),
+                  1.e1 * rebracketing_fac * v(iblock, pT, k, j, i),
```
Comment (brryan) [inline src/radiation/moments_source.cpp:369]: The original bracket window of 10 decades could cause precision issues in some cases, especially with the `itp` rootfinder.

File: tst/regression/mocmc_diffusion.gold [hunk 7]
```diff
@@ -1,25 +1,25 @@
-2.578814481287522261e-07
-8.539726882840010634e-11
-9.881445129638454866e-11
-2.880460357023297414e-08
-2.198153555894490595e-03
-6.512936231879706828e-01
-1.589897079949697579e-04
-2.487714603862677815e-02
-2.197674042014615206e-02
+2.578814481287529143e-07
```
Comment (brryan) [inline tst/regression/mocmc_diffusion.gold:1]: These upgolds are probably due to the changed numerics of the radiation source rootfind

File: tst/regression/regression_test.py [hunk 8]
```diff
@@ -17,7 +17,7 @@
 from subprocess import call
 import shutil
 import glob
-sys.path.append("../../external/parthenon/scripts/python/packages/parthenon_tools")
+sys.path.insert(0,"../../external/parthenon/scripts/python/packages/parthenon_tools")
```
Comment (brryan) [inline tst/regression/regression_test.py:20]: The `append`ed `parthenon_tools` would be ignored if a (potentially different) version of `parthenon_tools` was installed locally.

File: scripts/python/get_torus_fluxes.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+import os
+from subprocess import call, DEVNULL
+import glob
+import time
+from enum import Enum
+from phoebus_constants import cgs, scalefree
+import phoebus_utils
+from phoedf import phoedf
+
+def get_torus_fluxes(dfile):
```
Comment (Yurlungur) [inline scripts/python/get_torus_fluxes.py:33]: This is convenient. But why not just request phoebus to store and report them?

File: scripts/python/phoedf.py [hunk 10]
```diff
@@ -46,6 +46,14 @@ def __init__(self, filename, geomfile=None):
     self.Nx3 = self.MeshBlockSize[2]
 
     self.ScalarField = [self.NumBlocks, self.Nx3, self.Nx2, self.Nx1]
+    self.ThreeVectorField = [self.NumBlocks, 3, self.Nx3, self.Nx2, self.Nx1]
+    self.FourVectorField = [self.NumBlocks, 4, self.Nx3, self.Nx2, self.Nx1]
+    self.ThreeTensorField = [self.NumBlocks, 3, 3, self.Nx3, self.Nx2, self.Nx1]
+    self.FourTensorField = [self.NumBlocks, 4, 4, self.Nx3, self.Nx2, self.Nx1]
```
Comment (Yurlungur) [inline scripts/python/phoedf.py:52]: cute

File: scripts/python/phoedf.py (hunk 1 above)
Comment (Yurlungur) [reply scripts/python/phoedf.py:120]: @AstroBarker you may find some useful formulae in here for your Julia stuff

File: src/radiation/moments_source.cpp [hunk 11]
```diff
@@ -387,14 +403,14 @@ TaskStatus MomentFluidSourceImpl(T *rc, Real dt, bool update_fluid) {
               c.GetConTilPiFromPrim(J, cov_H, &con_tilPi);
             }
 
-            // Real JBB = d_opacity.EnergyDensityFromTemperature(T1, species_d[ispec]);
```
Comment (Yurlungur) [inline src/radiation/moments_source.cpp:390]: delete?

File: src/radiation/moments_source.cpp [hunk 12]
```diff
@@ -387,14 +403,14 @@ TaskStatus MomentFluidSourceImpl(T *rc, Real dt, bool update_fluid) {
...
             Real JBB = opacities.EnergyDensityFromTemperature(T1, species_d[ispec]);
-            // Real kappa = d_mean_opacity.RosselandMeanAbsorptionCoefficient(
```
Comment (Yurlungur) [inline src/radiation/moments_source.cpp:392]: delete?

File: tst/regression/mocmc_diffusion.gold (hunk 7 above)
Comment (Yurlungur) [reply tst/regression/mocmc_diffusion.gold:1]: do the answers still look right?

**Review by Yurlungur — APPROVED** [review]
Should also have @lroberts36 look at this I may have missed something in the rad

File: tst/regression/mocmc_diffusion.gold (hunk 7 above)
Comment (brryan) [reply tst/regression/mocmc_diffusion.gold:1]: yep, they differ at like the 10th decimal place or something.

File: scripts/python/get_torus_fluxes.py (hunk 9 above)
Comment (brryan) [reply scripts/python/get_torus_fluxes.py:33]: Yeah I think once we have all the fluxes we care about in logfiles we can go ahead and delete this file from the repo

File: src/radiation/moments_source.cpp (hunk 11 above)
Comment (brryan) [reply src/radiation/moments_source.cpp:390]: This PR actually does delete this :)

Comment (Yurlungur) [reply src/radiation/moments_source.cpp:390]: lol oops

File: src/pgen/torus.cpp [hunk 13]
```diff
@@ -427,7 +427,10 @@ void ProblemModifier(ParameterInput *pin) {
   }
 }
 
-void PostInitializationModifier(ParameterInput *pin, Mesh *pmesh) {
+void PostInitializationModifier(ParameterInput *pin, Mesh *pmesh, const bool is_restart) {
+  if (is_restart) {
+    return;
+  }
```
Comment (lroberts36) [inline src/pgen/torus.cpp:433]: Should `PostInitializationModifier` even be called for restarts?

File: src/radiation/moments.cpp [hunk 14]
```diff
@@ -831,11 +831,12 @@ TaskStatus CalculateGeometricSourceImpl(T *rc, T *rc_src) {
         Real dlnalp[ND];
         Real Gamma[ND][ND][ND];
         geom.GradLnAlpha(CellLocation::Cent, iblock, k, j, i, dlnalp);
-        geom.ConnectionCoefficient(CellLocation::Cent, iblock, k, j, i, Gamma);
+        Real dg[ND][ND][ND];
+        geom.MetricDerivative(CellLocation::Cent, k, j, i, dg);
+        Geometry::Utils::SetConnectionCoeffFromMetricDerivs(dg, Gamma);
```
Comment (lroberts36) [inline src/radiation/moments.cpp:836]: What is the reason for this change? I don't see `dg` being used below (but maybe I missed it)

File: src/radiation/moments.cpp (hunk 5 above)
Comment (lroberts36) [reply src/radiation/moments.cpp:869]: Good catch, I agree it makes more sense to just work with the lab frame quantities.

File: src/radiation/moments.cpp (hunk 14 above)
Comment (lroberts36) [reply src/radiation/moments.cpp:836]: nm, I see where it is used below

**Review by lroberts36 — APPROVED** [review]
This looks good to me

File: src/pgen/torus.cpp (hunk 13 above)
Comment (brryan) [reply src/pgen/torus.cpp]: Yeah you're right I don't see any reason to ever call `PostInitializationModifier` for restarts -- I removed the `is_restart` parameter and now just don't call this function at all on restarts.
